test-kotest/src/main/kotlin/io/micronaut/test/extensions/kotest/MicronautKotestExtension.kt

```diff
@@ -0,0 +1,96 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.test.extensions.kotest
+
+import io.kotest.core.extensions.ConstructorExtension
+import io.kotest.core.extensions.TestCaseExtension
+import io.kotest.core.listeners.TestListener
+import io.kotest.core.spec.Spec
+import io.kotest.core.test.TestCase
+import io.kotest.core.test.TestResult
+import io.micronaut.aop.InterceptedProxy
+import io.micronaut.test.annotation.MicronautTest
+import kotlin.reflect.KClass
+import kotlin.reflect.full.primaryConstructor
+
+object MicronautKotestExtension: TestListener, ConstructorExtension, TestCaseExtension {
+
+    override suspend fun intercept(
+        testCase: TestCase,
+        execute: suspend (TestCase) -> TestResult
+    ): TestResult {
+        val context = contexts[testCase.spec.javaClass.name]
+        return if(context != null && context.getSpecDefinition() == null) {
+            // It's a MicronautTest test where the bean doesn't exist
+            TestResult.Ignored
+        } else {
+            // Not a MicronautTest test or the bean exists
+            execute(testCase)
+        }
+    }
+
+    val contexts: MutableMap<String, MicronautKotestContext> = mutableMapOf()
+
+    override suspend fun beforeSpec(spec: Spec) {
+        contexts[spec.javaClass.name]?.beforeSpecClass(spec)
+    }
+
+    override suspend fun afterSpec(spec: Spec) {
+        contexts[spec.javaClass.name]?.afterSpecClass(spec)
+    }
+
+    override suspend fun beforeTest(testCase: TestCase) {
+        contexts[testCase.spec.javaClass.name]?.beforeTest(testCase)
+    }
+
+    override suspend fun afterTest(testCase: TestCase, result: TestResult) {
+        contexts[testCase.spec.javaClass.name]?.afterTest(testCase)
+    }
+    
+    
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : Spec> instantiate(clazz: KClass<T>): Spec? {
+        val constructor = clazz.primaryConstructor
+        val testClass: Class<Any> = clazz.java as Class<Any>
+        val micronautTest = testClass.annotations.filterIsInstance<MicronautTest>().firstOrNull()
+
+        return if (micronautTest == null) {
+            null
+        } else {
+            val createBean = constructor != null && constructor.parameters.isNotEmpty()
+            val context = MicronautKotestContext(testClass, micronautTest, createBean)
+            contexts[testClass.name] = context
+            if (createBean) {
+                context.bean
+            } else {
+                null
+            }
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T> Spec.getMock(obj: T): T {
+        return if (obj is InterceptedProxy<*>) {
+            obj.interceptedTarget() as T
+        } else {
+            obj
+        }
+    }
+}
+
+@Deprecated(message = "MicornautKotlinTestExtension is deprecated", replaceWith = ReplaceWith("MicronautKotlinTestExtension"))
```

Comment:
Since the KotestExtension is new, this deprecation typealias is unnecessary (and misleading)

Reply:
I didn't remove this typealias as a way to not break code that already works. We can remove it/change the deprecation message if @graemerocher or @jameskleeh think it's better

Reply:
The deprecation I commented is a workaround for a typo that seems to have existed in the MicronautKotlinTest extension, pointing to the correctly spelled version of that one.

Since the KotestExtension is new and living in a separate package, there will be no code using io.micronaut.test.kotest.MicornautKotlinTestExtension, but if at all, it uses the one from the kotlintest package.

Reply:
The deprecation I commented is a workaround for a typo that seems to have existed in the MicronautKotlinTest extension, pointing to the correctly spelled version of that one.

Since the KotestExtension is new and living in a separate package, there will be no code using io.micronaut.test.kotest.MicornautKotlinTestExtension, but if at all, it uses the one from the kotlintest package.